Two blank lines after an import should be reduced to one

kastkeepitjumpinlikekangaroos opened the pull request:
### Description

Hey! Trying to contribute for the first time to the project so let me know if I'm missing anything :) 

I've implemented changes to address this issue https://github.com/psf/black/issues/2020 where there's some inconsistency around how many lines are added after an import. I've added 2 test cases showing that if there is more than 1 blank line between the import and the next line that it is now being reduced to one. One test case was failing because it had in its expected outout 2 lines between the import and following statement, so I think modifying the test case now that it should always be one makes sense.

### Checklist - did you ...

- [x] Add an entry in `CHANGES.md` if necessary?
- [x] Add / update tests if necessary?
- [x] Add new / update outdated documentation?

JelleZijlstra commented on src/black/lines.py:
```diff
@@ -605,7 +605,16 @@ def _maybe_empty_lines(self, current_line: Line) -> tuple[int, int]:  # noqa: C9
             first_leaf = current_line.leaves[0]
             before = first_leaf.prefix.count("\n")
-            before = min(before, max_allowed)
+            before = (
```
This change should be applied only to the preview style; you'll need to add a feature to the enum and use Mode here to decide whether to make the change. Look for other recent changes that changed the formatting style for examples.

kastkeepitjumpinlikekangaroos replied:
Thanks!! after taking a look at some of the other PRs for preview styles I've added a new feature to the enum, docs, etc and added another condition here to make sure it's in the Mode. I updated the change log to have this change in the preview section as well. Based on the changes in the other PRs I believe I've updated everything necessary but please let me know if I'm missing anything else!

kastkeepitjumpinlikekangaroos commented:
I got an email about this job failure but looks like all the checks are passing on the PR https://github.com/psf/black/actions/runs/11377698063/job/31652281800 is this something that needs to be addressed?

JelleZijlstra commented:
Interesting, I suspect some change to how GitHub handles artifacts. Not sure why it doesn't show up here.

JelleZijlstra commented on tests/data/cases/preview_import_line_collapse.py:
```diff
@@ -0,0 +1,13 @@
+# flags: --preview
```
Could you add some more test cases, e.g. a comment line right after the import? Not every test case needs to be in its own file.

kastkeepitjumpinlikekangaroos replied:
sure thing! I added a few more test cases to this file and the other one, let me know if you think there should be any other test cases covered!

kastkeepitjumpinlikekangaroos replied:
also modified the logic to collapse to one line with comments as well since removing that condition didn't affect any existing test cases

kastkeepitjumpinlikekangaroos commented:
> I got an email about this job failure but looks like all the checks are passing on the PR https://github.com/psf/black/actions/runs/11377698063/job/31652281800 is this something that needs to be addressed?

I got this same error again on the rerun of the workflows https://github.com/psf/black/actions/runs/11483099904/job/31957735676 and seems to be because there's no artifact in this workflow https://github.com/psf/black/actions/runs/11412528080 with the file name `.pr-comment.json`. If I'm understanding correctly it looks like this is because these tasks which call `comment-body` (which produces `.pr-comment.json` https://github.com/psf/black/blob/main/scripts/diff_shades_gha_helper.py#L167) only execute when the matrix mode is `preview-changes` and the above `diff-shades` workflow has a matrix mode of `analysis`

https://github.com/psf/black/blob/main/.github/workflows/diff_shades.yml#L130:
```
      - name: Generate summary file (PR only)
        if: github.event_name == 'pull_request' && matrix.mode == 'preview-changes'
        run: >
          python helper.py comment-body
          ${{ matrix.baseline-analysis }} ${{ matrix.target-analysis }}
          ${{ matrix.baseline-sha }} ${{ matrix.target-sha }}
          ${{ github.event.pull_request.number }}

      - name: Upload summary file (PR only)
        if: github.event_name == 'pull_request' && matrix.mode == 'preview-changes'
        uses: actions/upload-artifact@v3
        with:
          name: .pr-comment.json
          path: .pr-comment.json
```

kastkeepitjumpinlikekangaroos commented:
hey @JelleZijlstra ! :) just bumping this PR - do the changes look good or are there other modifications you'd like to see?

hauntsaninja reviewed (commented):
Thanks, this looks reasonable

One nit: could you move the logic down next to https://github.com/psf/black/blob/c98fc0c128e1a79a776b62046acaa28685a31089/src/black/lines.py#L674-L680 and also add a test case for when imports are within a function?

kastkeepitjumpinlikekangaroos commented:
> Thanks, this looks reasonable
> 
> One nit: could you move the logic down next to
> 
> https://github.com/psf/black/blob/c98fc0c128e1a79a776b62046acaa28685a31089/src/black/lines.py#L674-L680
> 
> and also add a test case for when imports are within a function?

@hauntsaninja for sure, let me know if where I moved the logic is what you were thinking!

also moved all the case cases to a single file, added a test case for imports inside a function, and expanded some existing test cases

JelleZijlstra commented on CHANGES.md:
```diff
@@ -18,6 +18,7 @@
 <!-- Changes that affect Black's preview style -->
 
 - Remove parentheses around sole list items (#4312)
+- Collapse multiple lines into 1 after import (#4489)
```
```suggestion
- Collapse multiple empty lines after an import into one (#4489)
```

JelleZijlstra commented on docs/the_black_code_style/future_style.md:
```diff
@@ -41,6 +41,8 @@ Currently, the following features are included in the preview style:
 - `remove_lone_list_item_parens`: remove redundant parentheses around lone list items
   (depends on unstable `hug_parens_with_braces_and_square_brackets` feature in some
   cases)
+- `always_one_newline_after_import`: Always force one newline after import statements
+  except for when the line after the import is a comment or an import statement
```
```suggestion
- `always_one_newline_after_import`: Always force one blank line after import 
  statements, except when the line after the import is a comment or an import statement
```

JelleZijlstra reviewed (approved):
Thanks, just some wordsmithing

webartifex commented:
@steve-mavens I am not that familiar with the internals in `black`, just a happy user.

I tried to make it work for our code base by adapting `isort` and `ruff` but could'nt make it.

> "another thing that could happen is ruff format updates to match 2026 black style"

I would also vote for this!

AdmiralNemo commented:
> we strongly recommend trying out the --preview style sometime around November or December

...

> I feel a bit bad that I did not run black with the --preview flag all year in 2025; otherwise, I would have brought it up.

To me and my team, the point of `black` is that we can stop worrying about formatting.  Everyone just runs `black` before committing, and it takes away an entire class of discussion points in code reviews.   We were completely taken by surprise when CI checks started failing because of code that didn't even change.  It seems a bit disingenuous to request that we now have to spend time every year, during one of the busiest times for software engineering professionals, to pay attention to something we specifically adopted so we could stop paying attention to it.  I do not feel bad, I feel betrayed.
